Explicitly mark as unsupported instead of crashing mapping a parameter record type defined not in a namespace

bricknerb opened the pull request:
Before this change we crash in this case when trying to use the outer type.
After this change we diagnose a TODO.
Will add the missing support for this in a separate PR.

Part of #5533.

jonmeow commented on toolchain/check/testdata/interop/cpp/function/class.carbon:
```diff
@@ -165,6 +165,47 @@ fn F() {
+};
+
+auto foo1(P::C) -> void;
+auto foo2(P) -> void;
```
In each of these, you have this `foo2` with a call. However, is it necessary for the test (it's not inside the namespace)? If not, I'd suggest removing it; I'm not clear it's testing distinct behavior. If so, please add a comment clarifying what it's doing.

jonmeow replied:
Or, if you mean to test behavior of a type that contains another type, maybe that should be a separate test to clarify what you're trying to test?

bricknerb replied:
Thanks, done.
I've changed this PR to be made of:
* First commit: Just adds tests that don't diagnose for `class` and `struct` and wrongly treat the outer record as a namespace. This can be a separate PR.
* The rest fixes that and adds the tests that would have crashed without the fix.

jonmeow replied:
You've split tests. But I don't think you're addressing my question of what the test is achieving.

That is, why is "Pass outer class after passing inner class" an interesting thing to test, particularly when including all the SemIR for both calls? I agree that's what it does, but what is the test going to be testing that is meaningfully different? (if you just want to test passing the outer class that has a nested class, why not only test that and have less IR? why is "after" important?)

jonmeow replied:
To be clear, I'm asking you to *comment your code*.

bricknerb committed:
Add tests for passing a param by value of a record type defined in another record type.
The code assumes the outer record is a namespace.

jonmeow commented on toolchain/check/testdata/interop/cpp/function/class.carbon:
```diff
@@ -165,6 +165,86 @@ fn F() {
+// Pass outer class after passing inner class
+// ============================================================================
+
+// --- inner_and_outer_value_param_type.h
+
+class O {
+ public:
+  class C {};
+};
+
+auto foo1(O::C) -> void;
+auto foo2(O) -> void;
```
Can this same header be shared, rather than copying it?

bricknerb replied:
Done.

jonmeow commented on toolchain/check/testdata/interop/cpp/function/class.carbon:
```diff
@@ -165,6 +165,86 @@ fn F() {
+import Cpp library "inner_and_outer_value_param_type.h";
+
+fn F() {
+  //@dump-sem-ir-begin
```
Does the SemIR need to be captured for the first call? Will it differ from the above test?

bricknerb replied:
Done.

jonmeow commented on toolchain/check/testdata/interop/cpp/function/class.carbon:
```diff
@@ -165,6 +165,86 @@ fn F() {
+  // CHECK:STDERR:   ^~~~~~~~
+  // CHECK:STDERR:
+  Cpp.foo1({});
+  Cpp.foo2({});
```
If you're just trying to verify that `Cpp.O` is a type, maybe:

```suggestion
  //@dump-sem-ir-begin
  // Verify `O` is a valid type.
  var x: Cpp.O;
```

bricknerb replied:
Trying this out actually revealed another bug here, which I've sent a fix for in #5789.
I think it's better to fix that first.

bricknerb replied:
Done.

bricknerb committed:
Explicitly mark as unsupported instead of crashing mapping a parameter record type defined not in a namespace
Before this change we crash in this case.
After this change we diagnose a TODO.
Will add the missing support for this in a separate PR.

Part of #5533.

bricknerb committed:
Don't output the SemIR generated by the call to the function that triggers importing the record.

bricknerb committed:
Remove unused `foo2()`

bricknerb committed:
Reuse the `definition_in_*_value_param_type` header instead of duplicating it.

bricknerb committed:
Rename files to be more consistent

bricknerb commented:
I believe I've addressed all comments.
Had to rebase and ended up basically recreating this change, so some history is lost. :(

jonmeow commented on toolchain/check/testdata/interop/cpp/function/class.carbon:
```diff
@@ -211,6 +211,58 @@ fn F() {
+  // CHECK:STDERR:
+  Cpp.foo({});
+  //@dump-sem-ir-begin
+  var x: Cpp.O;
```
Since the second check (`x: Cpp.O`) is no longer calling, I think it'd make sense to re-combine the two tests now. It just hadn't been clear to me what you intended to test before: if you were testing the type was usable post-`Cpp.foo`, or if you were testing that a type containing a type could be used at all.

bricknerb replied:
Done.

jonmeow reviewed (approved):
Basically LG, see comment

bricknerb committed:
Merge definition in outer definition tests
